File: src/main/java/org/example/math_geometry/MagicSquaresInGrid.kt
package org.example.math_geometry


fun numMagicSquaresInside(grid: Array<IntArray>): Int {
    val ROWS = grid.size; val COLS = grid[0].size

    // returns 1 if square is magic else 0
    fun magic(r: Int, c: Int): Int {
        // Ensure 1-9
        val values = mutableSetOf<Int>()
        for (i in r..<r + 3) {
            for (j in c..<c + 3) {
                if (grid[i][j] in values || grid[i][j] !in 1..9) {
                    return 0
                }
                values.add(grid[i][j])
            }
        }

        // Row Sum
        for (i in r..<r + 3){
            if (grid[i].slice(c..<c + 3).sum() != 15) {
                return 0
            }
        }

        // Col Sum
        for (i in c..<c + 3) {
            if (grid[r][i] + grid[r+1][i] + grid[r+2][i] != 15) {
                return 0
            }
        }

        // Diagonal Sum
        if (grid[r][c] + grid[r+1][c+1] + grid[r+2][c+2] != 15 ||
            grid[r][c+2] + grid[r+1][c+1] + grid[r+2][c] != 15) {
            return 0
        }

        return 1
    }

    var count = 0
    for (r in 0 until ROWS - 2) {
        for (c in 0 until COLS - 2) {
            count += magic(r, c)
        }
    }

    return count
}

File: src/main/java/org/example/math_geometry/MagicSquaresInGrid.py
from typing import List


class Solution:
    # T: O(R * C), M: O(1)
    def numMagicSquaresInside(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])

        def magic(r, c):  # returns 1 for magic else 0
            # Ensure 1 - 9
            values = set()
            for i in range(r, r + 3):
                for j in range(c, c + 3):
                    if grid[i][j] in values or not (0 < grid[i][j] < 10):
                        return 0
                    values.add(grid[i][j])

            # row sum
            for i in range(r, r + 3):
                if sum(grid[i][c:c + 3]) != 15:
                    return 0

            # col sum
            for i in range(c, c + 3):
                if grid[r][i] + grid[r + 1][i] + grid[r + 2][i] != 15:
                    return 0

            # diagonal sum
            if (
                grid[r][c] + grid[r+1][c+1] + grid[r+2][c+2] != 15 or
                grid[r][c+2] + grid[r+1][c+1] + grid[r+2][c] != 15
            ):
                return 0
            return 1

        res = 0
        # only check for row, col positions from where a magic square can be made
        for r in range(ROWS - 2):
            for c in range(COLS - 2):
                res += magic(r, c)

        return res